### PR TITLE
fixed a race condition in NodePublishVolume/NodeUnpublishVolume

### PR DESCRIPTION
**Pull request by gman0**
Adds a mutex and a map with VolumeId as a key. Permits NodePublishVolume
and NodeUnpublishVolume functions to continue only when the map doesn't
contain that VolumeId, otherwise returns an error.

Rationale:

NodePublishVolume's SafeFormatAndMount may take quite a long time. This would
in turn trigger CO's timeout which would re-run NodePublishVolume again.

On a Linux system, SafeFormatAndMount calls fsck, mkfs.ext4 and mount.
Running those commands repeatedly and simultaneously on the same volume
may well break it.

As a follow-up question, is rbdplugin even supposed to format and mount the volume? It's a block device after all.

### CODE REVIEW

File: pkg/rbd/nodeserver.go [hunk 1]
```diff
@@ -38,11 +39,46 @@ type nodeServer struct {
...
+
+	delete(pendingVols, volID)
+}
+
 func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
 	targetPath := req.GetTargetPath()
+	volumeID := req.GetVolumeId()
+
+	if isVolumePending(volumeID) {
+		return nil, fmt.Errorf("rbd: NodePublishVolume for %s is pending", volumeID)
```
Comment (rootfs) [inline pkg/rbd/nodeserver.go]: return a proper error to allow retry 
https://github.com/container-storage-interface/spec/blob/master/spec.md#nodepublishvolume-errors

File: pkg/rbd/nodeserver.go [hunk 2]
```diff
@@ -38,11 +39,46 @@ type nodeServer struct {
 	*csicommon.DefaultNodeServer
 }
 
+var (
```
Comment (rootfs) [inline pkg/rbd/nodeserver.go]: reuse keymutex https://github.com/kubernetes/kubernetes/blob/master/pkg/util/keymutex/

**Comment by gman0** [comment]
As for your second comment, a try-lock functionality is required - it should fail immediately, not block till the currently running NodePublishVolume finishes. For that, I've submitted https://github.com/kubernetes/kubernetes/pull/60212 let's see if and when it gets merged :D

**Comment by gman0** [comment]
I believe this is related https://github.com/kubernetes-csi/external-provisioner/issues/63

**Comment by sbezverk** [comment]
@gman0 @rootfs it is really strange that you run into this situation because NodePublishVolume calls must be blocking calls. So it is suppsosed to just sit and wait until it is either finished or timed out and failed. We are going to review this flow from outside perspective. external-provisioner/attacher and k8s bits. I would suggest to hold with this changes until core components are cleaned up. We also will change volume naming to be deterministic in the next quarter..

**Comment by pohly** [comment]
@sbezverk are you sure that NodePublishVolume is a blocking call? My understanding of gRPC is that each request is processed independently, in its own goroutine. It's the responsibility of the implementer to protect against race conditions.

Let me propose a different PR, one where mutex locking inside the process is used to serialize calls - see PR #92

**Comment by pohly** [comment]
I think a more complete version of this was merged in PR #92 and this one can be closed now?

**Pull request closed without merging**